process/estimator/executionResultInclusionEstimator.go

```diff
@@ -0,0 +1,191 @@
+	currentHdrTsMs uint64,
+) (allowed int) {
+	allowed = 0
+
+	if len(pending) == 0 {
+		return allowed
+	}
+
+	var tBase uint64
+	// lastNotarised is nil if genesis.
+	if lastNotarised == nil {
+		tBase = convertMsToNs(erie.GenesisTimeMs)
+	} else {
+		tBase = convertMsToNs(lastNotarised.HeaderTimeMs)
+	}
+
+	currentHdrTsNs := convertMsToNs(currentHdrTsMs)
+
+	// accumulated execution time in ns (1 gas = 1ns)
+	estimatedTime := uint64(0)
+	for i, executionResultMeta := range pending {
+		var previousExecutionResultMeta *ExecutionResultMetaData
+		if i > 0 {
+			previousExecutionResultMeta = &pending[i-1]
+		}
+		ok := erie.checkSanity(executionResultMeta, previousExecutionResultMeta, lastNotarised, currentHdrTsNs)
+		if !ok {
+			return i
+		}
+
+		overflow, currentEstimatedTime := bits.Mul64(executionResultMeta.GasUsed, erie.tGas)
+		if overflow != 0 {
+			log.Debug("ExecutionResultInclusionEstimator: overflow detected in currentEstimatedTime",
+				"currentEstimatedTime", currentEstimatedTime,
+				"gasUsed", executionResultMeta.GasUsed,
+				"tGas", erie.tGas,
+			)
+			return i
+		}
+
+		estimatedTime, overflow = bits.Add64(estimatedTime, currentEstimatedTime, 0)
+		if overflow != 0 {
+			log.Debug("ExecutionResultInclusionEstimator: overflow detected in block transactions time estimation",
+				"estimatedTime", estimatedTime,
+				"currentEstimatedTime", currentEstimatedTime)
+			return i
+		}
+
+		// Apply safety margin
+		overflow, estimatedTimeWithMargin := bits.Mul64(estimatedTime, erie.cfg.SafetyMargin)
+		if overflow != 0 {
+			log.Debug("ExecutionResultInclusionEstimator: overflow detected in estimated time with margin",
+				"estimatedTime", estimatedTime,
+				"safetyMargin", erie.cfg.SafetyMargin)
+			return i
+		}
+		estimatedTimeWithMargin /= 100
+
+		tDone, overflow := bits.Add64(tBase, estimatedTimeWithMargin, 0)
+		if overflow != 0 {
+			log.Debug("ExecutionResultInclusionEstimator: overflow detected in total estimated time",
+				"tBase", tBase,
+				"estimatedTimeWithMargin", estimatedTimeWithMargin)
+			return i
+		}
+
+		// check for time cap reached, cannot include current pending item or anything after
+		if tDone > currentHdrTsNs {
+			log.Debug("ExecutionResultInclusionEstimator: estimated time exceeds current header timestamp",
+				"tDone", tDone,
+				"currentHdrTsNs", currentHdrTsNs)
+			return i
+		}
+
+		// check for number of results cap reached, including current pending item. MaxResultsPerBlock = 0 means no cap.
+		if erie.cfg.MaxResultsPerBlock != 0 && uint64(i+1) >= erie.cfg.MaxResultsPerBlock {
+			log.Debug("ExecutionResultInclusionEstimator: reached MaxResultsPerBlock cap",
+				"maxResultsPerBlock", erie.cfg.MaxResultsPerBlock,
+				"currentIndex", i+1)
+			return i + 1
+		}
+	}
+
+	// If we reach here, all pending items are safe to include
+	return len(pending)
+}
+
+func (erie ExecutionResultInclusionEstimator) checkSanity(currentExecutionResultMeta ExecutionResultMetaData,
```

Comment:
use pointer receiver in all the places 
`(erie *ExecutionResultInclusionEstimator)`